<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Annotating Variants}
-->

# ![](/images/icons/help.gif)Annotating Variants 

The VariantAnnotation package has facilities for reading in all or portions 
of Variant Call Format (VCF) files. Structural location information can be 
determined as well as amino acid coding changes for non-synonymous variants. 
Consequences of the coding changes can be investigated with the SIFT and 
PolyPhen database packages.


* [Sample Workflow](#sample-workflow) 
* [Installation and Use](#install-and-use)
* [Exploring Package Content](#exploring-package-content)
* [Resources](#resources)

<h2 id="sample-workflow">Sample Workflow</h2>
 
Performed with Bioconductor 3.0 and R >= 3.1;
VariantAnnotation 1.11.35.

This workflow annotates variants found in the Transient Receptor Potential 
Vanilloid (TRPV) gene family on chromosome 17. The VCF file is available in 
the cgdv17 data package and contains Complete Genomics data for population 
type CEU.

```{r echo=FALSE}
suppressPackageStartupMessages(library(VariantAnnotation))
suppressPackageStartupMessages(library(cgdv17))
suppressPackageStartupMessages(library(org.Hs.eg.db))
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg19))
suppressPackageStartupMessages(library(PolyPhen.Hsapiens.dbSNP131))
```

```{r}
library(VariantAnnotation)
library(cgdv17)
file <- system.file("vcf", "NA06985_17.vcf.gz", package = "cgdv17")
 
## Explore the file header with scanVcfHeader
hdr <- scanVcfHeader(file)
 
info(hdr) 
 
geno(hdr) 
```
     
Convert the gene symbols to gene ids compatible with the 
TxDb.Hsapiens.UCSC.hg19.knownGene annotations. The annotaions 
are used to define the TRPV ranges that will be extracted from
the VCF file.

```{r}
## get entrez ids from gene symbols
library(org.Hs.eg.db)
genesym <- c("TRPV1", "TRPV2", "TRPV3")
geneid <- select(org.Hs.eg.db, keys=genesym, keytype="SYMBOL",
                 columns="ENTREZID")
geneid
```
     
Load the annotation package.

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
```

Modify the seqlevels (chromosomes) in the txdb to match
those in the VCF file. This step is necessary because we want
to use ranges from the txdb to extract a subset from the VCF.

```{r}
renameSeqlevels(txdb, gsub("chr", "", seqlevels(txdb)))
keepSeqlevels(txdb, "17")
```

Create a list of transcripts by gene:

```{r}
txbygene = transcriptsBy(txdb, "gene")
```

Create the gene ranges for the TRPV genes

```{r}
gnrng <- unlist(range(txbygene[geneid$ENTREZID]), use.names=FALSE)
names(gnrng) <- geneid$SYMBOL
```

A ScanVcfParam object is used to retrieve data subsets. This object
can specify genomic coordinates (ranges) or individual VCF elements.
Extractions of ranges (vs fields) requires a tabix index.
See ?indexTabix for details.

```{r}
param <- ScanVcfParam(which = gnrng, info = "DP", geno = c("GT", "cPd"))
param
 
## Extract the TRPV ranges from the VCF file 
vcf <- readVcf(file, "hg19", param)
## Inspect the VCF object with the 'fixed', 'info' and 'geno' accessors
vcf
 
head(fixed(vcf))

geno(vcf)
```

To find the structural location of the variants, we'll use the 
locateVariants function with the TxDb.Hsapiens.UCSC.hg19.knownGene 
package loaded eariler.

## Use the 'region' argument to define the region
## of interest. See ?locateVariants for details.
cds <- locateVariants(vcf, txdb, CodingVariants())
five <- locateVariants(vcf, txdb, FiveUTRVariants())
splice <- locateVariants(vcf, txdb, SpliceSiteVariants())
intron <- locateVariants(vcf, txdb, IntronVariants())

```{r}     
all <- locateVariants(vcf, txdb, AllVariants())
```
     
Each row in cds represents a variant-transcript match so multiple rows
per variant are possible. If we are interested in gene-centric questions
the data can be summarized by gene regardless of transcript.

```{r}     
## Did any variants match more than one gene?
table(sapply(split(mcols(all)$GENEID, mcols(all)$QUERYID), 
      function(x) length(unique(x)) > 1))
 
## Summarize the number of variants by gene:
idx <- sapply(split(mcols(all)$QUERYID, mcols(all)$GENEID), unique)
sapply(idx, length)
 
## Summarize variant location by gene:
sapply(names(idx), 
    function(nm) {
        d <- all[mcols(all)$GENEID %in% nm, c("QUERYID", "LOCATION")]
        table(mcols(d)$LOCATION[duplicated(d) == FALSE])
    })
```
     
Amino acid coding for non-synonymous variants can be computed
with the function predictCoding. The BSgenome.Hsapiens.UCSC.hg19 
package is used as the source of the reference alleles. Variant 
alleles are provided by the user.

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
seqlevelsStyle(vcf) <- "UCSC"
seqlevelsStyle(txdb) <- "UCSC"
aa <- predictCoding(vcf, txdb, Hsapiens)
```
     
predictCoding returns results for coding variants only. As with 
locateVariants, the output has one row per variant-transcript match
so multiple rows per variant are possible.

```{r}     
## Did any variants match more than one gene?
table(sapply(split(mcols(aa)$GENEID, mcols(aa)$QUERYID), 
        function(x) length(unique(x)) > 1))

## Summarize the number of variants by gene:
idx <- sapply(split(mcols(aa)$QUERYID, mcols(aa)$GENEID, drop=TRUE), unique)
sapply(idx, length)

## Summarize variant consequence by gene:
sapply(names(idx), 
       function(nm) {
           d <- aa[mcols(aa)$GENEID %in% nm, c("QUERYID","CONSEQUENCE")]
           table(mcols(d)$CONSEQUENCE[duplicated(d) == FALSE])
       })
```
    
The variants 'not translated' are explained by the warnings thrown when
predictCoding was called. Variants that have a missing varAllele or have an
'N' in the varAllele are not translated. If the varAllele substitution had
resulted in a frameshift the consequence would be 'frameshift'. See
?predictCoding for details.
    
The SIFT.Hsapiens.dbSNP132 and PolyPhen.Hsapiens.dbSNP131 packages 
provide predictions of how damaging amino acid coding changes may
be to protein structure and function. Both packages search on rsid.
    
The pre-computed predictions in the SIFT and PolyPhen packages are based 
on specific gene models. SIFT is based on Ensembl and PolyPhen on UCSC 
Known Gene. The TranscriptDb we used to identify coding variants was 
from UCSC Known Gene so we will use PolyPhen for predictions.

```{r}    
## Load the PolyPhen package and explore the available keys and columns
library(PolyPhen.Hsapiens.dbSNP131)
keys <- keys(PolyPhen.Hsapiens.dbSNP131)
cols <- columns(PolyPhen.Hsapiens.dbSNP131)
## column descriptions are found at ?PolyPhenDbColumns
cols

## Get the rsids for the non-synonymous variants from the
## predictCoding results
rsid <- unique(names(aa)[mcols(aa)$CONSEQUENCE == "nonsynonymous"]) 

## Retrieve predictions for non-synonymous variants. Two of the six variants 
## are found in the PolyPhen database. 
select(PolyPhen.Hsapiens.dbSNP131, keys=rsid, 
       columns=c("AA1", "AA2", "PREDICTION"))
```

<p class="back_to_top">[ <a href="#top">Back to top</a> ]</p>


<h2 id="install-and-use">Installation and Use</h2>

Follow [installation instructions](/install/) to start using these
packages.  To install VariantAnnotation use 

```{r eval=FALSE}
library(BiocInstaller) 
biocLite("VariantAnnotation")
```

Package installation is required only once per R installation. View a
full list of available
[software](/packages/release/bioc/)
and 
[annotation](/packages/release/data/annotation/)
packages.

To use the `VariantAnnotation`, evaluate the commands

```{r eval=FALSE}
library(VariantAnnotation)
````    

These commands are required once in each R session.

<p class="back_to_top">[ <a href="#top">Back to top</a> ]</p>

<h2 id="exploring-package-content">Exploring Package Content</h2>

Packages have extensive help pages, and include vignettes highlighting
common use cases. The help pages and vignettes are available from
within R. After loading a package, use syntax like

    help(package="VariantAnnotation")
    ?predictCoding

to obtain an overview of help on the `VariantAnnotation` package, and 
the `predictCoding` function. View the package vignette with

```{r eval=FALSE}
browseVignettes(package="VariantAnnotation")
```

To view vignettes providing a more comprehensive introduction to
package functionality use

```{r eval=FALSE}
help.start()
```


<p class="back_to_top">[ <a href="#top">Back to top</a> ]</p>

```{r}
sessionInfo()
```

<p class="back_to_top">[ <a href="#top">Back to top</a> ]</p>
